libraries/AP_OSD/AP_OSD_ParamSetting.cpp

```diff
@@ -0,0 +1,418 @@
+            }
+            incr = MAX(1, powf(10, digits - 2));
+            max = powf(10, digits + 1);
+            debug("Guessing range for value %d as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_INT32:
+        {
+            AP_Int32* p = (AP_Int32*)_param;
+            min = -1;
+            uint8_t digits = 0;
+            for (int32_t int32p = p->get(); int32p > 0; int32p /= 10) {
+                digits++;
+            }
+            incr = MAX(1, powf(10, digits - 2));
+            max = powf(10, digits + 1);
+            debug("Guessing range for value %d as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_FLOAT:
+        {
+            AP_Float* p = (AP_Float*)_param;
+
+            uint8_t digits = 0;
+            for (float floatp = p->get(); floatp > 1.0f; floatp /= 10) {
+                digits++;
+            }
+            float floatp = p->get();
+            if (digits < 1) {
+                if (!is_zero(floatp)) {
+                    incr = floatp / 100.0f; // move in 1% increments
+                } else {
+                    incr = 0.01f; // move in absolute 1% increments
+                }
+                max = 1.0;
+                min = 0.0f;
+            } else {
+                if (!is_zero(floatp)) {
+                    incr = floatp / 100.0f; // move in 1% increments
+                } else {
+                    incr = MAX(1, powf(10, digits - 2));
+                }
+                max = powf(10, digits + 1);
+                min = 0.0f;
+            }
+            debug("Guessing range for value %f as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_VECTOR3F:
+        case AP_PARAM_NONE:
+        case AP_PARAM_GROUP:
+            break;
+        }
+
+        if (!_param_min.configured() || force) {
+            _param_min = min;
+        }
+        if (!_param_max.configured() || force) {
+            _param_max = max;
+        }
+        if (!_param_incr.configured() || force) {
+            _param_incr = incr;
+        }
+    }
+}
+
+bool AP_OSD_ParamSetting::set_from_metadata()
+{
+    // check for statically configured setting metadata
+    if (_type > 0 && _type < OSD_PARAM_NUM_TYPES && _param_metadata[_type - 1].values_max > 0) {
+        _param_incr = _param_metadata[_type - 1].increment;
+        _param_min = _param_metadata[_type - 1].min_value;
+        _param_max = _param_metadata[_type - 1].max_value;
+        return true;
+    }
+    return false;
+}
+
+// modify the selected parameter values
+void AP_OSD_ParamSetting::save_as_new()
+{
+    _param_group.save();
+    _param_key.save();
+    _param_idx.save();
+    // the user has configured the range and increment, but the parameter
+    // is no longer valid so reset these to guessed values
+    guess_ranges(true);
+    if (_param_min.configured()) {
+        _param_min.save();
```

Comment:
prefer to avoid save() if we can to save storage

Reply:
Bitmask used to check whether parameter has been modified